File: adventofcode/year2018/Day03.cs
using adventofcode.common;
using adventofcode.common.grid;
using adventofcode.common.range;
using adventofcode.common.util;
using System.Text.RegularExpressions;

namespace adventofcode.year2018;

public class Day03 : Solution
{
    private Dictionary<int, Box2D> _squares;
    
    public Day03(string year, string day) : base(year, day)
    {
        _squares = new Dictionary<int, Box2D>();
        foreach (var line in LoadInputAsLines())
        {
            var match = Regex.Match(line, @"#(\d+) @ (\d+),(\d+): (\d+)x(\d+)");
            var topLeft = new Point2D(Convert.ToInt32(match.Groups[2].Value), Convert.ToInt32(match.Groups[3].Value));
            var bottomRight = topLeft + new Point2D(Convert.ToInt32(match.Groups[4].Value), Convert.ToInt32(match.Groups[5].Value));
            _squares.Add(Convert.ToInt32(match.Groups[1].Value), new Box2D(topLeft, bottomRight));
        }
    }

    public override string PartOne()
    {
        var overlaps = new HashSet<Point2D>();
        foreach (var combination in IterTools<Box2D>.Combination(_squares.Values.ToList(), 2))
        {
            var intersection = combination[0].Intersect(combination[1]);
            if (intersection is not null)
            {
                for (int x = 0; x < intersection.Width(); x++)
                {
                    for (int y = 0; y < intersection.Height(); y++)
                    {
                        overlaps.Add(new Point2D(intersection.TopLeft().X() + x, intersection.TopLeft().Y() + y));
                    }
                }
            }
        }

        return Convert.ToString(overlaps.Count);
    }

    public override string PartTwo()
    {
        var squareID = -1;
        foreach (var a in _squares)
        {
            var overlap = false;
            foreach (var b in _squares)
            {
                if (a.Key == b.Key) { continue; }
                if (a.Value.Intersect(b.Value) is not null)
                {
                    overlap = true;
                    break;
                }
            }

            if (!overlap)
            {
                squareID = a.Key;
                break;
            }
        }

        return Convert.ToString(squareID);
    }
}

File: adventofcode/year2025/day03.py
from __future__ import annotations
from adventofcode.common import Solution


def largest_joltage(battery_bank: str, num_batteries: int) -> str:
    joltage = []
    largest_index = 0
    max_index = len(battery_bank) - num_batteries
    for b in range(num_batteries):
        largest = '0'
        i = largest_index
        while i <= max_index:
            if battery_bank[i] > largest:
                largest = battery_bank[i]
                largest_index = i
            i += 1
        joltage.append(largest)
        max_index += 1
        largest_index += 1
    return "".join(joltage)


class Day03(Solution):
    def __init__(self, year: str, day: str):
        super().__init__(year, day)
        self._input = self._load_input_as_lines()

    def part_one(self):
        count = 0
        for battery_bank in self._input:
            joltage = largest_joltage(battery_bank, 2)
            print(f"{battery_bank} -> {joltage}")
            count += int(joltage)
        return count

    def part_two(self):
        count = 0
        for battery_bank in self._input:
            joltage = largest_joltage(battery_bank, 12)
            print(f"{battery_bank} -> {joltage}")
            count += int(joltage)
        return count
